text/0000-canonical-metadata.md

```diff
@@ -0,0 +1,673 @@
+
+## Code generation
+Finally, metadata can facilitate low code development environment by allowing developer to declaratively define component. This metadata can be used to generate component stubs and the developer can add the specific business logic into the stubs thus reducing the amount of manually authored boiler plate code.
+Metadata can also be transformed to generate components in various eco-systems like LWC, native web components etc. This would require adding an additional transformation of the metadata to another format like the custom-elements-json in the case of web components.
+
+# Current State of Art
+## Web Components community
+There is no well established standard or format for declaratively defining a web component. [custom-elements-json](https://github.com/webcomponents/custom-elements-json) is making an attempt to define a standard. It is in a very early stage. However there is a ecosystem developing around this standard. 
+[web-component-analyzer](https://www.npmjs.com/package/web-component-analyzer) can analyze web components written in vanilla javascript and other popular web component libraries. It also supports custom-elements-json as an output format.
+
+## LWC Platform
+Currently, LWC platform gathers metadata as part of the compilation process. This metadata shape and implementation is currently private. This RFC is an attempt to make the metadata shape public as a first step. This [document](https://salesforce.quip.com/DdncANrJA0ko)<sup>*</sup> captures the information collected and the metadata shape currently gathered. Here is a summary of problems with the current implementation:
+* Replicated type system in java and javascript that is maintained manually
+* Dependency(reference) analysis is done outside of metadata gathering
+* Change management to accomodate new metadata is cumbersome because the implementation is spread across two repos
+
+# Detailed Design
+
+## Format
+Metadata shape of the various objects we wish to analyse will be represented using in [JSON Schema](https://json-schema.org/) format. This is also augmented with a TypeScript definition of the shape.
+
+[JSON](https://tools.ietf.org/html/rfc7159) is the IETF approved standard for representing objects. [JSON Schema](https://json-schema.org/) is on the path to become an IETF approved standard for describing JSON.
+
+This format was also influenced by standardization choices made with in Salesforce UI platform.
+
+## Entities that will be analyzed
+These are the entities that will be analzed and metadata gathered about:
+* HTML Template file
+    * Custom element references
+        * Tag name
+        * Attributes
+        * Slot content - default and named slots
+    * Static resource
+        * Image urls
+        * Javascript resource
+        * CSS resource
+    * Dynamic components
+    * Template directives
```

Comment:
directives should probably fold under custom element references section. Alternatively, since we already expose bindingAST, which contains such information, perhaps the directives section can be left out all together